### PR TITLE
Clarify documentation about when device uses APNs production/ development device tokens

### PR DESCRIPTION
**Pull request by ben-xD**
## TLDR

Clarify what distribution method (app store connect, ad hoc, enterprise, development)/ build configuration (release, debug, profile) causes the device to use APNs production vs. development. The PR also includes a small token auth documentation improvement, replacing an undefined symbol for a magic string (`"ABLY_API_KEY"`) instead. 

I also added a fix for images showing on pub.dev, by using the `raw.githubusercontent.com` domain, the images will show up in pub.dev.

## More detail: what are the differences?:
Distribution methods available: 
Screenshot 2021-08-29 at 07 37 33

Build configurations available:
CleanShot 2021-08-29 at 07 37 22

### CODE REVIEW

File: PushNotifications.md [hunk 1]
```diff
@@ -62,7 +62,7 @@ To get push notifications setup in your own app, read [Setting up your own app](
...
+       - When running your application via Xcode or your machine (Android Studio, command line), your application runs in either debug, profile or release mode. In all cases, your application will use the sandbox/ development APNs environment. When distributing your app in the App Store, Ad Hoc or through App Store Connect, it will always use the production environment. If distributing through `Development`, the sandbox/ development APNs environment is used. Keep in mind your distribution method when checking the `Use APNS sandbox environment?` checkbox in a Ably application on the Ably dashboard (notifications tab). Ensure your application connects to an Ably application configured to use production APNs if the application is being distributed publicly.
```
Comment (QuintinWillison) [inline PushNotifications.md:65]: Please include a space either side of `/` in sentences. Also, 'Ably application' needs to be preceded by 'an', not 'a'.

```suggestion
       - When running your application via Xcode or your machine (Android Studio, command line), your application runs in either debug, profile or release mode. In all cases, your application will use the sandbox/ development APNs environment. When distributing your app in the App Store, Ad Hoc or through App Store Connect, it will always use the production environment. If distributing through `Development`, the sandbox / development APNs environment is used. Keep in mind your distribution method when checking the `Use APNS sandbox environment?` checkbox in an Ably application on the Ably dashboard (notifications tab). Ensure your application connects to an Ably application configured to use production APNs if the application is being distributed publicly.
```

File: PushNotifications.md [hunk 2]
```diff
@@ -302,6 +302,16 @@ On iOS, to show alert notifications, you need to request provisional permission
 
 On Android, you can use logcat built into Android Studio or [pidcat](https://github.com/JakeWharton/pidcat) to view the logs.
 
+### I have confirmed messages are not being received by the device (no logs of being cancelled), but no errors are thrown when I sent the message.
```
Comment (QuintinWillison) [inline PushNotifications.md:305]: Tense correction. Plus no need for headings to have full stop endings.

```suggestion
### I have confirmed messages are not being received by the device (no logs of being cancelled), but no errors are thrown when I send the message
```

File: PushNotifications.md [hunk 3]
```diff
@@ -302,6 +302,16 @@ On iOS, to show alert notifications, you need to request provisional permission
...
+
+Sending an Ably message with a push payload will succeed if the message is successfully delivered to Ably, including delivery guarantees for channel subscriptions. However, push notification messages may not be delivered to specific devices if there is an error returned by APNs/ FCM. You should retrieve the device registration using a Push Admin or the Ably Dashboard. For example, if your application has a bundle ID which does not match APNs certificate generated from developer.apple.com,  you would get `DeviceTokenNotForTopic`, on the Ably dashboard, this would look like:
```
Comment (QuintinWillison) [inline PushNotifications.md:307]: Spacing around '/', please. Also, perhaps worth making the Apple domain a link.

```suggestion
Sending an Ably message with a push payload will succeed if the message is successfully delivered to Ably, including delivery guarantees for channel subscriptions. However, push notification messages may not be delivered to specific devices if there is an error returned by APNs / FCM. You should retrieve the device registration using a Push Admin or the Ably Dashboard. For example, if your application has a bundle ID which does not match APNs certificate generated from [developer.apple.com](https://developer.apple.com/account/resources/identifiers/list), you would get `DeviceTokenNotForTopic`, on the Ably dashboard, this would look like:
```

File: PushNotifications.md [hunk 4]
```diff
@@ -314,24 +324,21 @@ Ensure your Android app contains the Firebase configuration `android/app/google-
...
 
-For example, for `BadDeviceToken`:
 > The specified device token was bad. Verify that the request contains a valid token and that the token matches the environment.
 
-When running a debug application, the sandbox/ development APNs server is used. Make sure to use an application with "Use APNS sandbox environment" enabled in the Ably dashboard (push notification tab).  Do not try to change the `.entitlements` file to `production`, but this does not make the debug application use the production APNs server. For more information about this limitation, see [How do I make my debug app version receive production push notifications on iOS?](https://stackoverflow.com/a/46118155/7365866).
+When running a debug application, the sandbox/ development APNs server is used. Make sure to use an application with `Use APNS sandbox environment?` enabled in the Ably dashboard (push notification tab). Changing the `aps-environment` value in `.entitlements` file to `production` does not make the debug application use the production APNs server.
```
Comment (QuintinWillison) [inline PushNotifications.md:333]: I failed to spot the lack of uniform spacing around the '/' in my previous review.

```suggestion
When running a debug application, the sandbox / development APNs server is used. Make sure to use an application with `Use APNS sandbox environment?` enabled in the Ably dashboard (push notification tab). Changing the `aps-environment` value in the `.entitlements` file to `production` does not make the debug application use the production APNs server.
```

File: PushNotifications.md [hunk 5]
```diff
@@ -314,24 +324,21 @@ Ensure your Android app contains the Firebase configuration `android/app/google-
...
 
->**Debug** builds will always get *sandbox* APNS tokens.
+>**Development/ Sandbox APNs:** Local builds through Xcode/ Android Studio/ command line will always get *sandbox* APNs tokens. Apps distributed through `Development` methods will also get *sandbox* APNs tokens.
```
Comment (QuintinWillison) [inline PushNotifications.md:335]:
```suggestion
> **Development / Sandbox APNs:** Local builds through Xcode / Android Studio/ command line will always get *sandbox* APNs tokens. Apps distributed through `Development` methods will also get *sandbox* APNs tokens.
```

File: PushNotifications.md [hunk 6]
```diff
@@ -314,24 +324,21 @@ Ensure your Android app contains the Firebase configuration `android/app/google-
...
 >
->**Release** builds (ad-hoc or app store) will always get *production* APNS tokens.
+>**Production APNs:** Apps distributed through App Store Connect (TestFlight and App Store), Ad Hoc and Enterprise distribution methods will always get *production* APNs tokens.
```
Comment (QuintinWillison) [inline PushNotifications.md:337]:
```suggestion
> **Production APNs:** Apps distributed through App Store Connect (TestFlight and App Store), Ad Hoc and Enterprise distribution methods will always get *production* APNs tokens.
```

File: PushNotifications.md [hunk 7]
```diff
@@ -314,24 +324,21 @@ Ensure your Android app contains the Firebase configuration `android/app/google-
...
 
 For more information, take a look at [What are the possible reasons to get APNs responses BadDeviceToken or Unregistered?](https://stackoverflow.com/questions/42511476/what-are-the-possible-reasons-to-get-apns-responses-baddevicetoken-or-unregister).
 
-### Android only: When I look in ably.com's dashboard, i see "InvalidRegistration"
+### Android: When retrieving a device registration using a Push Admin or using the Ably dashboard, the device push state is is `InvalidRegistration`?
```
Comment (QuintinWillison) [inline PushNotifications.md:341]: Sentence not constructed as a question.

```suggestion
### Android: When retrieving a device registration using a Push Admin or using the Ably dashboard, the device push state is is `InvalidRegistration`
```

**Review by QuintinWillison — CHANGES_REQUESTED** [review]

**Review by QuintinWillison — APPROVED** [review]
I notice you've evolved this a lot since my last comments. I've not really got the time right now to re-review at the level of detail I would like to, so I think let's take the balanced view that "more / improved" docs is always a good / better thing.
